runtime/src/util/mapmap.rs

```diff
@@ -0,0 +1,103 @@
+use cid::Cid;
+use fvm_ipld_blockstore::Blockstore;
+use fvm_ipld_hamt::{Error, BytesKey};
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+use std::collections::BTreeMap;
+
+use crate::{make_empty_map, make_map_with_root_and_bitwidth, Keyer, Map};
+
+// MapMap stores multiple values per key in a Hamt of Hamts
+// Every element stored has a primary and secondary key
+pub struct MapMap<'a, BS, V> 
+{
+    outer: Map<'a, BS, Cid>,
+    inner_bitwidth: u32, 
+    cache: BTreeMap<Vec<u8>, Map<'a, BS, V>>,
+}
+impl<'a, BS, V> MapMap<'a, BS, V>
+where
+    BS: Blockstore,
+    V: Serialize + DeserializeOwned + Clone + std::cmp::PartialEq,
+{
+
+    pub fn new(bs: &'a BS, outer_bitwidth: u32, inner_bitwidth: u32) -> Self {
+        MapMap {
+            outer: make_empty_map(bs, outer_bitwidth),
+            inner_bitwidth,
+            cache: BTreeMap::<Vec<u8>, Map<BS, V>>::new(),
+        }
+    }
+
+    pub fn from_root(bs: &'a BS, cid: &Cid, outer_bitwidth: u32, inner_bitwidth:u32 ) -> Result<Self, Error> {
+        Ok(MapMap{
+            outer: make_map_with_root_and_bitwidth(cid, bs, outer_bitwidth)?,
+            inner_bitwidth,
+            cache: BTreeMap::<Vec<u8>, Map<BS, V>>::new(),
+        })
+    }
+
+    pub fn flush(&mut self) -> Result<Cid, Error> {
+        for (k,in_map) in self.cache.iter_mut() {
+            if in_map.is_empty() {
+                self.outer.delete(&BytesKey(k.to_vec()))?;
+            } else {
+                let new_in_root = in_map.flush()?;
+                self.outer.set(BytesKey(k.to_vec()), new_in_root)?;
+            }
+        }
+        self.outer.flush()
+    }
+
+    fn load_inner_map<K1>(& mut self, k: K1)  -> Result<& mut Map<'a, BS, V>, Error> 
+    where
+        K1: Keyer + std::fmt::Debug,
+    {
+        let in_map_thunk = || -> Result<Map<BS, V>, Error> {
```

Comment:
I am quite worried about this whole method.  It took a lot of time for me to get to this structure.  Without using the entry api I was totally blocked by the borrow checker.

This all seems to be working about as expected given test behavior.  However when I instrument this thunk with a print statment it looks like it gets called even in the case that the entry already exists in the BTree map and the result is not used.  This should still have strictly better performance because it only flushes once even if it redundantly reads twice, but it would be best to eliminate redundant reads too.

Reply:
You couldn't get it working with an `if`? I'm afraid I don't have the expertise to help, maybe flag down a rust expert to help (independent of merging this)

Reply:
Ok I figured this out.  I was making confused about some very basic things.  Looking into the implementation of the entry api this is actually really simple.  or_insert just takes in data and matches against it so it makes perfect sense the thunk was getting evaluated.  We can just do the match explicitly to get the behavior we want.